pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,527 @@
+	"github.com/cilium/cilium/pkg/datapath/state"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/ip"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+// DevicesControllerCell registers a controller that subscribes to network devices
+// and routes via netlink and populates the devices and routes devices.
+var DevicesControllerCell = cell.Module(
+	"datapath-devices-controller",
+	"Populates the device and route tables from netlink",
+
+	cell.Invoke(registerDevicesController),
+	cell.Provide(newDeviceManager),
+)
+
+var (
+	// batchingDuration is the amount of time to wait for more
+	// addr/route/link updates before processing the batch.
+	batchingDuration = 100 * time.Millisecond
+
+	excludedDevicePrefixes = []string{
+		defaults.HostDevice,
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+	}
+
+	// Route filter to look at all routing tables.
+	routeFilter = netlink.Route{
+		Table: unix.RT_TABLE_UNSPEC,
+	}
+	routeFilterMask = netlink.RT_FILTER_TABLE
+)
+
+type DevicesConfig struct {
+	Devices []string
+}
+
+type devicesControllerParams struct {
+	cell.In
+
+	Log         logrus.FieldLogger
+	DB          statedb.DB
+	DeviceTable statedb.Table[*state.Device]
+	RouteTable  statedb.Table[*state.Route]
+	Config      DevicesConfig
+
+	// NetNS is an optional network namespace handle. Used by tests to restrict the detection
+	// to specific namespace.
+	NetNS *netns.NsHandle `optional:"true"`
+}
+
+func registerDevicesController(lc hive.Lifecycle, p devicesControllerParams) {
+	dc := &devicesController{
+		devicesControllerParams: p,
+		links:                   map[int]netlink.Link{},
+		filter:                  deviceFilter(p.Config.Devices),
+	}
+	dc.ctx, dc.cancel = context.WithCancel(context.Background())
+	lc.Append(dc)
+}
+
+type devicesController struct {
+	devicesControllerParams
+	filter         deviceFilter
+	l3DevSupported bool
+	handle         *netlink.Handle
+	ctx            context.Context
+	cancel         context.CancelFunc
+	links          map[int]netlink.Link
+}
+
+func (dc *devicesController) Start(hive.HookContext) error {
+	dc.l3DevSupported = probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
+
+	// Construct a handle that uses the configured network namespace.
+	netns := netns.None()
+	if dc.NetNS != nil {
+		netns = *dc.NetNS
+	}
+	var err error
+	dc.handle, err = netlink.NewHandleAt(netns)
+	if err != nil {
+		return fmt.Errorf("NewHandleAt failed: %w", err)
+	}
+
+	// Subscribe to address updates to find out about link address changes
+	// to update the device's addresses.
+	addrUpdates := make(chan netlink.AddrUpdate, 16)
+	opts := netlink.AddrSubscribeOptions{
+		Namespace:    dc.NetNS,
+		ListExisting: false,
+	}
+	if err := netlink.AddrSubscribeWithOptions(addrUpdates, dc.ctx.Done(), opts); err != nil {
```

Comment:
I believe in the past we've seen issues where the netlink can be unreliabe, should we try to reestablish the nl connection/sub if the channel gets closed?

Reply:
Oh that's a bummer. Would require fair bit of additional complexity to resubscribe as it needs to relist from scratch in order to make sure no update has been missed.

Do you have any more info about the unreliability?

Reply:
netlink man page has this to say: 
```
      Netlink is not a reliable protocol.  It tries its best to deliver
       a message to its destination(s), but may drop messages when an
       out-of-memory condition or other error occurs.  For reliable
       transfer the sender can request an acknowledgement from the
       receiver by setting the NLM_F_ACK flag.  An acknowledgement is an
       NLMSG_ERROR packet with the error field set to 0.  The
       application must generate acknowledgements for received messages
       itself.  The kernel tries to send an NLMSG_ERROR message for
       every failed packet.  A user process should follow this
       convention too.

       However, reliable transmissions from kernel to user are
       impossible in any case.  The kernel can't send a netlink message
       if the socket buffer is full: the message will be dropped and the
       kernel and the user-space process will no longer have the same
       view of kernel state.  It is up to the application to detect when
       this happens (via the ENOBUFS error returned by [recvmsg(2)](https://man7.org/linux/man-pages/man2/recvmsg.2.html)) and
       resynchronize.
```

The go `netlink` library doesn't do the ACKs so this is indeed is pretty unreliable and may miss updates.
Not sure in what scenarios the netlink socket would get closed, but wouldn't be surprised it can happen
in OOM situations.

I'll need to think a bit how we can make this reliable...

Reply:
Yeah, correct.

A timeout was added by a contributor to the[ nl library](https://github.com/cilium/cilium/issues/14710) to solve issues where the socket would hang indefinitely (I don't think it was clear what the actual underlying issue was, potentially just a bug in netlink?).

Could we propose/add the feature ourselves?

Reply:
The recvfrom hang has been fixed 1 year ago: https://github.com/vishvananda/netlink/pull/793. Unfortunately the netlink library seems to not be maintained well anymore. I'll likely look into actively maintained alternative libraries that would also allow cleanly distinguishing between initial listing and updates.